ads/google/a4a/utils.js

```diff
@@ -293,3 +310,83 @@ export function getCorrelator(win, opt_cid) {
   }
   return win.ampAdPageCorrelator;
 }
+
+/**
+ * Gets a relatively terse and controlled-format string version of the user
+ * agent.
+ *
+ * @param {!Window} win
+ * @return {string}
+ */
+export function userAgentString(win) {
+  const platform = platformFor(win);
+  let browser;
+  let os;
+  let version;
+  if (platform.isIos()) {
+    os = 'ios';
+  } else if (platform.isAndroid()) {
+    os = 'android';
+  } else {
+    os = 'other_os';
+  }
+  if (platform.isSafari()) {
+    browser = 'safari';
+  } else if (platform.isChrome()) {
+    browser = 'chrome';
+  } else {
+    browser = 'other_browser';
+  }
+  version = platform.getMajorVersion();
+  return `${browser}_${version}_on_${os}`;
+}
+
+/**
+ * Creates or reinitializes a lifecycle reporter for Google ad network
+ * implementations.
+ *
+ * @param {!../../../extensions/amp-a4a/0.1/amp-a4a.AmpA4A} a4aElement
+ * @return {!./performance.GoogleAdLifecycleReporter}
+ */
+export function googleLifecycleReporterFactory(a4aElement) {
+  const reporter =
+      /** @type {!./performance.GoogleAdLifecycleReporter} */
+      (getLifecycleReporter(a4aElement, 'a4a', undefined,
+          a4aElement.element.getAttribute('data-amp-slot-index')));
+  const slotId = reporter.getSlotId();
+  reporter.setPingVariable('v_h', 'VIEWPORT_HEIGHT');
+  reporter.setPingVariable('s_t', 'SCROLL_TOP');
+  const eid = a4aElement.element.getAttribute(EXPERIMENT_ATTRIBUTE);
+  if (eid) {
+    reporter.setPingVariable('e', eid);
+  }
+  const type = a4aElement.element.getAttribute('type');
+  if (type) {
+    reporter.setPingVariable(`adt.${slotId}`, type);
+  }
+  reporter.setPingVariable('ua', userAgentString(a4aElement.win));
+  return reporter;
+}
+
+/**
+ * Sets reportable variables from ad response headers.
+ *
+ * @param {!FetchResponseHeaders} headers
+ * @param {!./performance.GoogleAdLifecycleReporter} reporter
+ */
+export function setGoogleLifecycleVarsFromHeaders(headers, reporter) {
+  const qqid = headers.get(QQID_HEADER);
+  if (qqid) {
+    reporter.setPingVariable(`qqid.${reporter.getSlotId()}`, qqid);
+  }
+  // This is duplicated from the amp-a4a.js implementation.  It needs to be
+  // defined there because it's an implementation detail of that module, but
+  // we want to report it to Google b/c we're interested in how rendering mode
+  // affects Google ads.  However, we can't directly reference a variable
+  // in extensions/ from here.
+  const renderingTypeHeader = 'X-AmpAdRender';
```

Comment:
It's not a question of exporting from A4A, but of dependency restrictions:

> gulp dep-check
> [14:48:19] ads/google/a4a/utils.js must not depend on extensions/amp-a4a/0.1/amp-a4a.js

I could whitelist, but that seems a bit extreme for a single constant.  I agree that this is not ideal, though.  The *real* problem is that we need a utils / shared library location for stuff in `extensions/`.  But we don't have that at the moment, and I'd rather not try to set that up in this PR.  :-)